Do not allow to reuse previous credentials in case of inter-server secret

azat opened the pull request:
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not allow to reuse previous credentials in case of inter-server secret (Before INSERT via Buffer/Kafka to Distributed table with interserver secret configured for that cluster, may re-use previously set user for that connection)

Fixes: #13156

azat commented:
Cc: @vitlibar

azat commented:
>Docs check — Found 1 errors

e18af442d47a2febbebaaf052f9d4a776a17c5d9

>Functional stateless tests (debug) — fail: 1, passed: 3363, skipped: 19

`2010_lc_native` - #29153

azat commented:
@alexey-milovidov maybe you can take a look? (or ping @vitlibar somehow / or at least assign the PR to him)

azat commented:
>Docs check — Found 1 errors

Unrelated, and already fixed in upstream/master

>Functional stateless tests (debug) — fail: 1, passed: 3363, skipped: 19

#29153

@vitlibar can you please take a look?

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -1212,13 +1212,23 @@ void TCPHandler::receiveQuery()
         {
-            LOG_DEBUG(log, "User (initial): {}", client_info.initial_user);
-            session->authenticate(AlwaysAllowCredentials{client_info.initial_user}, client_info.initial_address);
+            user = "default";
```
For example `Buffer` inserts into its underlying table using the global context, and the global context isn't always the same as the context of the user named `default`. So maybe in this case when we get a secondary query here with an empty initial user we should assume it's the global context.

And assuming that if `user.empty()` we should call here something like
`session->resetAuthentication()` instead of calling
`session->authenticate(AlwaysAllowCredentials{"default"})`.

azat replied:
>For example Buffer inserts into its underlying table using the global context, and the global context isn't always the same as the context of the user named default. 

Agree (that's why I wrote "at least" in the commit message :) )

> So maybe in this case when we get a secondary query here with an empty initial user we should assume it's the global context.

Indeed.

> And assuming that if user.empty() we should call here something like
session->resetAuthentication() instead of calling
session->authenticate(AlwaysAllowCredentials{"default"}).

Yep, great idea - done.

azat committed:
Add a test for INSERT w/o user in interserver mode

v2: ensure that the test fails with the version w/o fix
v3: force connect by modifying config and reload it
v4: add comments

azat commented:
>[Fast test — fail: 1, passed: 2912, skipped: 498. Other tests won't be run before fast test will be fixed.](https://clickhouse-test-reports.s3.yandex.net/29060/1127bcd65ca98f5249cabace6fa77391a28fc591/fast_test.html#fail1)

```
2021-09-28 01:43:49 2024_merge_regexp_assert:                                               [ FAIL ] - return code: 81
2021-09-28 01:43:49 Expected server error code: 47 but got: 81 (query: SELECT a FROM merge(REGEXP('\0'), '^t$'); -- { serverError 47 }).
2021-09-28 01:43:49 Received exception from server (version 21.11.1):
2021-09-28 01:43:49 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: Database test_xlj25w doesn't exist. (UNKNOWN_DATABASE)
2021-09-28 01:43:49 (query: SELECT a FROM merge(REGEXP('\0'), '^t$'); -- { serverError 47 })
```

https://github.com/ClickHouse/ClickHouse/pull/29461

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=29060) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=29060) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=29060) it.</sub>

vitlibar commented on src/Interpreters/Session.cpp:
```diff
@@ -326,6 +326,12 @@ void Session::authenticate(const Credentials & credentials_, const Poco::Net::So
+void Session::resetAuthentication()
+{
+    user_id.reset();
+    *prepared_client_info = {};
```
`user` should also be reset, not only `user_id`. And it seems there can be some fields in `prepared_client_info` which should not be reset, so I think it's better to reset only authentication-related fields in `prepared_client_info`. And please check if `prepared_client_info` is not null before.

azat replied:
Ok

azat committed:
tests/integration/helpers: add query_id for query()

azat commented:
@alesapin Looks like CI hanged?

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -1178,13 +1178,22 @@ void TCPHandler::receiveQuery()
         {
-            LOG_DEBUG(log, "User (initial): {}", client_info.initial_user);
+            LOG_DEBUG(log, "User (no user, interserver mode)");
+            session->resetAuthentication();
```
Alternative solution: it seems to solve the issue it's enough to just recreate the session with

```
session = std::make_unique<Session>(server.context(), ClientInfo::Interface::TCP_INTERSERVER);
```

before the line `ClientInfo client_info = session->getClientInfo()` in this function. This alternative solution appears to be much shorter and more logical than reseting authentication.

azat replied:
But this will loose all other fields, that is filled on Hello, i.e. `ClientInfo.client_*` fields, and others

vitlibar replied:
https://github.com/ClickHouse/ClickHouse/blob/acfa949586808bbc76b8e3423856601d80e6f7ad/src/Server/TCPHandler.cpp#L1137 will replace those fields anyway.

azat replied:
Then it is good idea, thanks, done!

azat replied:
It is still required to cleanup some fields in `ClientInfo` explicitly - #34263

azat committed:
Do not allow to reuse previous credentials in case of inter-server secret

Before this patch INSERT via Buffer/Kafka may re-use previously set user
for that connection, while this is not correct, it should reset the
user, and use global context.

Note, before [1] there was a fallback to default user, but that code had
been removed, and now it got back.

  [1]: 0159c74f217ec764060c480819e3ccc9d5a99a63 ("Secure inter-cluster query execution (with initial_user as current query user) [v3]")

Also note, that context for Buffer table (and others) cannot be changed,
since they don't have any user only profile.

I've tested this patch manually using the following:

    create table dist (key Int) engine=Distributed(test_cluster_two_shards_secure, default, data, key);
    create table buffer (key Int) engine=Buffer(default, dist, 1, 0, 0, 0, 0, 0, 0);
    create table data (key Int) engine=Memory();

    # to start the connection with readonly user
    $ clickhouse-client --user readonly -q 'select * from dist'
    $ clickhouse-client -q 'insert into buffer values (1)'
    # before this patch this produces errors like:
    # 2021.09.27 23:46:48.384920 [ 19474 ] {} <Error> default.dist.DirectoryMonitor: Code: 164. DB::Exception: Received from 127.0.0.2:9000. DB::Exception: readonly: Cannot execute query in readonly mode. Stack trace:

v2: reset the authentication instead of using default user (as suggested by @vitlibar)
v3: reset Session::user and introduce ClientInfo::resetAuthentication (as suggested by @vitlibar)
v4: reset the session every time in interserver mode (suggested by @vitlibar)

azat committed:
Fix SessionLog::addLoginSuccess() for ClientInfo::Interface::TCP_INTERSERVER

In this case there there can be no user.
CI report [1]:

    Address: 0x1f Access: read. Address not mapped to object.
    Stack trace: 0x24494b2f 0x244784c6 0x260fe1af 0x260ed672 0x260ddffa 0x26104880 0x2d62f3af 0x2d6300a1 0x2d926666 0x2d91fb7a 0x7f8bcea3c609 0x7f8bce963293
    3.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/string:0: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::operator=(std::__1::basic_string>
    3. ../src/Interpreters/SessionLog.cpp:216: DB::SessionLog::addLoginSuccess(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::__1::optional<std::__1::basic_string<char, st>
    4. ./obj-x86_64-linux-gnu/../src/Interpreters/Session.cpp:0: DB::Session::makeQueryContextImpl(DB::ClientInfo const*, DB::ClientInfo*) const @ 0x244784c6 in /usr/bin/clickhouse
    5.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:3299: std::__1::shared_ptr<DB::Context>::swap(std::__1::shared_ptr<DB::Context>&)
    5.2. inlined from ../contrib/libcxx/include/memory:3243: std::__1::shared_ptr<DB::Context>::operator=(std::__1::shared_ptr<DB::Context>&&)
    5. ../src/Server/TCPHandler.cpp:1208: DB::TCPHandler::receiveQuery() @ 0x260fe1af in /usr/bin/clickhouse
    6. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::receivePacket() @ 0x260ed672 in /usr/bin/clickhouse
    7. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x260ddffa in /usr/bin/clickhouse
    8. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:1643: DB::TCPHandler::run() @ 0x26104880 in /usr/bin/clickhouse

  [1]: https://clickhouse-test-reports.s3.yandex.net/29060/c087fc0ed5fbea133eb3dc3a64b8db93a81d0ece/integration_tests_flaky_check_(asan).html#fail1

azat commented:
Can someone add `force-test` label?
